nsls2forge_utils/auto_tick.py

```diff
@@ -0,0 +1,627 @@
+        ],
+    ),
+]
+
+BOT_RERUN_LABEL = {
+    "name": "bot-rerun",
+    "color": "#191970",
+    "description": (
+        "Apply this label if you want the bot to retry "
+        "issuing a particular pull-request"
+    ),
+}
+
+
+def run(feedstock_ctx, migrator, protocol='ssh', pull_request=True,
+        rerender=True, fork=False, organization='nsls-ii-forge', **kwargs):
+    """
+    For a given feedstock and migration run the migration
+
+    Parameters
+    ----------
+    feedstock_ctx: FeedstockContext
+        The node attributes
+    migrator: Migrator instance
+        The migrator to run on the feedstock
+    protocol : str, optional
+        The git protocol to use, defaults to ``ssh``
+    pull_request : bool, optional
+        If true issue pull request, defaults to true
+    rerender : bool
+        Whether to rerender
+    fork : bool
+        If true create a fork, defaults to true
+    organization: str, optional
+        GitHub organization to get repo from
+    gh : github3.GitHub instance, optional
+        Object for communicating with GitHub, if None build from $USERNAME
+        and $PASSWORD, defaults to None
+    kwargs: dict
+        The key word arguments to pass to the migrator
+
+    Returns
+    -------
+    migrate_return: MigrationUidTypedDict
+        The migration return dict used for tracking finished migrations
+    pr_json: dict
+        The PR json object for recreating the PR as needed
+    """
+    # get the repo
+    migrator.attrs = feedstock_ctx.attrs
+
+    branch_name = migrator.remote_branch(feedstock_ctx) + "_h" + uuid4().hex[0:6]
+
+    # TODO: run this in parallel
+    feedstock_dir, repo = get_repo(
+        ctx=migrator.ctx.session,
+        fctx=feedstock_ctx,
+        branch=branch_name,
+        organization=organization,
+        feedstock=feedstock_ctx.feedstock_name,
+        protocol=protocol,
+        pull_request=pull_request,
+        fork=fork,
+
+    )
+
+    recipe_dir = os.path.join(feedstock_dir, "recipe")
+
+    # migrate the feedstock
+    migrator.run_pre_piggyback_migrations(recipe_dir, feedstock_ctx.attrs, **kwargs)
+
+    # TODO - make a commit here if the repo changed
+
+    migrate_return = migrator.migrate(recipe_dir, feedstock_ctx.attrs, **kwargs)
+
+    if not migrate_return:
+        logger.critical(
+            "Failed to migrate %s, %s",
+            feedstock_ctx.package_name,
+            feedstock_ctx.attrs.get("bad"),
+        )
+        eval_cmd(f"rm -rf {feedstock_dir}")
+        return False, False
+
+    # TODO - commit main migration here
+
+    migrator.run_post_piggyback_migrations(recipe_dir, feedstock_ctx.attrs, **kwargs)
+
+    # TODO commit post migration here
+
+    # rerender, maybe
+    diffed_files = []
+    with indir(feedstock_dir), env.swap(RAISE_SUBPROC_ERROR=False):
+        msg = migrator.commit_message(feedstock_ctx)  # noqa
+        try:
+            eval_cmd("git add --all .")
+            eval_cmd(f"git commit -am '{msg}'")
```

Comment:
Hmm, probably calling external tools is unavoidable, unless one wants to invest time into using it via a library. If these commands are left, probably the above `rm -rf` is OK to keep.